### ISSUE TITLE
Support for Multiple Datasets and Domain-Specific Loss Calculation in Trainer

### ISSUE THREAD

**Issue by ghost** [huggingface/transformers#30725]
### Feature request

I am currently working on a project that involves sequence level distillation across multiple domains, requiring the handling of separate datasets for each domain within a single training loop. Specifically, the challenge involves integrating data from four distinct domains, computing loss individually per domain, and then aggregating these losses into a global loss measure that can guide the overall training process.

### Motivation

Ideally, the Trainer class would natively support the following features:

Multiple Dataset Handling: Ability to pass multiple datasets into the Trainer directly, with each dataset potentially representing a different domain.
Domain-Specific Loss Calculation: Support for defining and computing loss separately for each domain's dataset within the training loop and then integrating these losses into a global training objective.

### Your contribution

Currently, the Trainer class in the Transformers library supports passing a single dataset for training and evaluation. To handle multiple datasets or to calculate domain-specific losses, one must subclass the Trainer and override methods such as compute_loss, which complicates the implementation and integration of domain-specific training strategies.

**Comment by cw235** [comment]
Your feature request for enhancing the Trainer class in the Transformers library to support handling multiple datasets representing different domains and calculating domain-specific losses is indeed valuable for projects involving sequence level distillation across various domains. Here are some potential contributions that could address your requirements:

1. **Multiple Dataset Handling Support:**
   - Modify the Trainer class to accept multiple datasets representing different domains directly as input. This enhancement would streamline the integration of diverse data sources within a single training loop.

2. **Domain-Specific Loss Calculation Integration:**
   - Implement a mechanism within the Trainer class to define and compute losses separately for each domain's dataset during the training loop. This functionality would allow for domain-specific loss calculations and aggregation into a global training objective.

3. **Flexible Loss Aggregation Mechanism:**
   - Introduce a flexible mechanism for aggregating domain-specific losses into a global training objective. This feature would enable users to define custom aggregation strategies based on the specific requirements of their projects.

4. **Unified Training Interface:**
   - Enhance the Trainer class to provide a unified training interface that simplifies the implementation of domain-specific training strategies. This improvement would abstract away the complexity of subclassing and method overriding for users seeking to work with multiple datasets and domain-specific losses.

5. **Documentation and Examples:**
   - Update the documentation of the Trainer class to include clear explanations and examples demonstrating how to leverage the new features for handling multiple datasets and calculating domain-specific losses. Providing detailed guidance would facilitate the adoption of these functionalities by the community.

By incorporating these features and enhancements into the Trainer class of the Transformers library, users working on projects involving sequence level distillation across diverse domains would benefit from a more streamlined and efficient training process. Additionally, these improvements would contribute to the versatility and usability of the library for a broad range of applications requiring multi-domain data integration and domain-specific training strategies.

**Comment by jacklanda** [comment]
+1

**Comment by LysandreJik** [comment]
cc @SunMarc as well :)

**Comment by ucaslei** [comment]
+1

**Comment by fzyzcjy** [comment]
Hi, is there any updates? Thanks!

**Comment by csking101** [comment]
Hi, I am interested to take up this issue. Could someone give a few pointers to proceed?

**Comment by jacklanda** [comment]
> Hi, I am interested to take up this issue. Could someone give a few pointers to proceed?

https://sbert.net/docs/sentence_transformer/training_overview.html#multi-dataset-training

We can directly refer to the feature implemented by Sentence Transformers.

**Comment by sergiopaniego** [comment]
Hi @csking101! 

 I'm also interested in contributing to this issue. Are you open to collaborating? 😄

**Comment by csking101** [comment]
> Hi @csking101!
> 
> I'm also interested in contributing to this issue. Are you open to collaborating? 😄

Sure, that would be great! Where can we get in touch?

**Comment by sergiopaniego** [comment]
You may find me on Discord with the same handle as here

**Comment by sergiopaniego** [comment]
Working on [it](https://github.com/sergiopaniego/transformers/tree/multiple-dataset-loss-30725). 
Will open a PR soon 😄 !

**Comment by liuheng0111** [comment]
> Working on [it](https://github.com/sergiopaniego/transformers/tree/multiple-dataset-loss-30725). Will open a PR soon 😄 !

When will this feature be released? 
Will it support dynamic batch_size during training?

**Comment by zhanqiuzhang** [comment]
+1

**Comment by druvdub** [comment]
Hey @sergiopaniego do you still need a hand for this issue?

**Comment by sameerajashyam** [comment]
Hey @amyeroberts  @sergiopaniego, I’d like to work on this.

**Comment by Rocketknight1** [comment]
Still seeing pings on this so cc @sunmarc for visibility

**Comment by qgallouedec** [comment]
Hey all, sorry this has sat so long, we are catching up on the backlog. Status: the Trainer refactor @SunMarc mentioned when #40995 was closed has since landed (#43807, #44041), so the blocker is gone.

However, we're not going to take this one. The Trainer is already complex enough, and this is a handful of lines in a subclass: tag each domain dataset with a `domain_id`, concatenate, and mask the per-token loss per domain in `compute_loss`.

Self-contained example, runs as is:

```python
import torch
from datasets import Dataset, concatenate_datasets
from transformers import AutoModelForCausalLM, AutoTokenizer, Trainer, TrainingArguments


DOMAINS = ["legal", "medical", "code", "news"]
WEIGHTS = torch.tensor([1.0, 2.0, 0.5, 1.0])  # global loss = sum_d w_d * loss_d


class MultiDomainTrainer(Trainer):
    def __init__(self, *args, domain_weights, **kwargs):
        super().__init__(*args, **kwargs)
        self.domain_weights = domain_weights

    def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=None):
        domain_ids = inputs.pop("domain_ids")
        labels = inputs.pop("labels")
        logits = model(**inputs).logits

        # Per-token cross-entropy, shifted for causal LM
        shift_logits, shift_labels = logits[:, :-1], labels[:, 1:]
        per_token_loss = torch.nn.functional.cross_entropy(
            shift_logits.transpose(1, 2), shift_labels, reduction="none", ignore_index=-100
        )
        mask = shift_labels != -100

        # One loss per domain present in the batch, then the weighted sum
        loss = torch.zeros((), device=logits.device)
        for domain_id, weight in enumerate(self.domain_weights):
            domain_mask = mask & (domain_ids == domain_id).unsqueeze(1)
            num_tokens = domain_mask.sum()
            if num_tokens == 0:  # this domain isn't in this batch
                continue
            domain_loss = (per_token_loss * domain_mask).sum() / num_tokens
            loss = loss + weight * domain_loss
        return loss


def collate(examples, pad_token_id):
    max_len = max(len(example["input_ids"]) for example in examples)
    input_ids = torch.full((len(examples), max_len), pad_token_id)
    labels = torch.full((len(examples), max_len), -100)
    for i, example in enumerate(examples):
        ids = torch.tensor(example["input_ids"])
        input_ids[i, : len(ids)] = ids
        labels[i, : len(ids)] = ids
    return {
        "input_ids": input_ids,
        "attention_mask": (labels != -100).long(),
        "labels": labels,
        "domain_ids": torch.tensor([example["domain_id"] for example in examples]),
    }


model_id = "Qwen/Qwen3-0.6B"
tokenizer = AutoTokenizer.from_pretrained(model_id)
model = AutoModelForCausalLM.from_pretrained(model_id)

# One dataset per domain, tagged with its domain id, then concatenated
datasets = []
for domain_id, domain in enumerate(DOMAINS):
    dataset = Dataset.from_dict({"text": [f"a {domain} sentence number {i}" for i in range(32)]})
    dataset = dataset.map(lambda example: tokenizer(example["text"]), remove_columns="text")
    datasets.append(dataset.add_column("domain_id", [domain_id] * len(dataset)))
train_dataset = concatenate_datasets(datasets).shuffle(seed=0)

trainer = MultiDomainTrainer(
    model=model,
    args=TrainingArguments(output_dir="multi-domain", max_steps=5, remove_unused_columns=False),
    train_dataset=train_dataset,
    data_collator=lambda examples: collate(examples, tokenizer.pad_token_id),
    domain_weights=WEIGHTS,
)
trainer.train()
```

Two things worth knowing: `remove_unused_columns=False` is required or `domain_id` gets dropped before the collator, and batches mix domains, hence the `num_tokens == 0` skip. If you need one domain per batch instead, that is a custom sampler, not a loss change.

Closing as this is well covered by subclassing.
